source/slang/ir-constexpr.cpp

```diff
@@ -0,0 +1,531 @@
+                                        changedThisIteration = true;
+                                    }
+                                }
+                            }
+                        }
+
+                        // TODO: this currently only works if the callee has a definition,
+                        // because that is the only case where will generate IR values for
+                        // its parameter list. Otherwise we'd need to pull this information
+                        // from the function *type* for builtins.
+                        if(!calleeFunc->firstBlock)
+                            continue;
+
+                    }
+                }
+            }
+
+            if( bb != code->getFirstBlock() )
+            {
+                // A parameter in anything butr the first block is
+                // conceptually a phi node, which means its operands
+                // are the corresponding values from the terminating
+                // branch in a predecessor block.
+
+                UInt paramCounter = 0;
+                for( auto pp = bb->getFirstParam(); pp; pp = pp->getNextParam() )
+                {
+                    UInt paramIndex = paramCounter++;
+
+                    if(!isConstExpr(pp))
+                        continue;
+
+                    for(auto pred : bb->getPredecessors())
+                    {
+                        auto terminator = pred->getLastInst();
+                        if(terminator->op != kIROp_unconditionalBranch)
+                            continue;
+
+                        UInt operandIndex = paramIndex + 1;
+                        SLANG_RELEASE_ASSERT(operandIndex < terminator->getArgCount());
+
+                        auto operand = terminator->getArg(operandIndex);
+                        if( maybeMarkConstExpr(context, operand) )
+                        {
+                            changedThisIteration = true;
+                        }
+                    }
+                }
+            }
+
+        }
+
+        if( !changedThisIteration )
+            return anyChanges;
+
+        anyChanges = true;
+    }
+}
+
+// Validate use of `constexpr` within a function (in particular,
+// diagnose places where a value that must be contexpr depends
+// on a value that cannot be)
+void validateConstExpr(
+    PropagateConstExprContext*  context,
+    IRGlobalValueWithCode*      code)
+{
+    for( auto bb = code->getFirstBlock(); bb; bb = bb->getNextBlock() )
+    {
+        for( auto ii = bb->getFirstInst(); ii; ii = ii->getNextInst() )
+        {
+            if(isConstExpr(ii))
+            {
+                // For an instruction that must be `constexpr`, we need
+                // to ensure that its argumenst are all `constexpr`
+
+                UInt argCount = ii->getArgCount();
+                for( UInt aa = 0; aa < argCount; ++aa )
+                {
+                    auto arg = ii->getArg(aa);
+
+                    if( !isConstExpr(arg) )
```

Comment:
Where do you create the RateQualifiedType for IRDeclRef and IRWitnessTable operands? If they are having the wrong type without the rate qualifier, this check will fail, right?

Reply:
The short answer is that I don't handle those cases.

This constant-ness condition currently only applies to computation on simple values, and so things like decl-refs and witness tables in the IR don't enter into things (they mostly appear as operands to `specialize` instructions).

The one big gotcha I suppose would be field references, where the result should be constant as long as the base expression was constant, and so we want to treat the decl-ref operand as constant too.

The `isConstExpr` query there has special-case logic where certain opcodes are considered constant by default, so that we don't have to actually propagate rates into their types (e.g., a literal is always a constant), so I suppose decl-refs could be handled there.